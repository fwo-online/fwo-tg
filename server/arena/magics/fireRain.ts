import { AoeDmgMagic } from '../Constuructors/AoeDmgMagicConstructor';
import type GameService from '../GameService';
import type { Player } from '../PlayersService';

/**
 * Огненный дождь
 * Основное описание магии общее требовани есть в конструкторе
 */
class FireRain extends AoeDmgMagic {
  constructor() {
    super({
      name: 'fireRain',
      displayName: 'Огненный дождь',
      desc: 'Обрушивает на команду противника огненный дождь',
      cost: 18,
      baseExp: 8,
      costType: 'mp',
      lvl: 3,
      orderType: 'enemy',
      aoeType: 'team',
      magType: 'bad',
      chance: [92, 94, 95],
      effect: ['1d2', '1d2+2', '1d2+4'],
      dmgType: 'fire',
      profList: ['m'],
    });
  }

  getTargets(): Player[] {
    const { target, game } = this.params;

    if (!target.clan) {
      return [];
    }

    const targets = game.players.getPlayersByClan(target.clan?.id)
      .filter(({ alive }) => alive)
      .filter(({ id }) => id !== target.id);

    return targets;
  }

  /**
   * Основная функция запуска магии
   */
  run(initiator: Player, target: Player, game: GameService): void {
    const effect = this.effectVal();
    target.stats.down('hp', effect);

    const targets = this.getTargets();

    targets.forEach((target) => {
      this.runAoe(initiator, target, game);
    });
  }

  runAoe(initiator: Player, target: Player, game: GameService) {
    const effect = this.effectVal({ initiator, target, game });
    target.stats.down('hp', effect);

    this.status.expArr.push({
      initiator,
      target,
      val: effect,
      hp: target.stats.val('hp'),
    });
  }
}

export default new FireRain();
