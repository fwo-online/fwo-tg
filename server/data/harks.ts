export const harksList = ['str', 'dex', 'wis', 'int', 'con'] as const;

export type Hark = typeof harksList[number];

export type HarksLvl = Record<Hark, number>;

export const harksData: Record<Hark, {
  name: string;
  descr: string;
}> = {
  str: {
    name: 'Сила',
    descr: '● *40%* влияет на вероятность пробоя \n● Увеличивает физический урон',
  },
  dex: {
    name: 'Ловкость',
    descr: '● *90%* влияет на вероятность пробоя \n● *40%* влияет на вероятность защиты \n● *60%* влияет на востановление энергии',
  },
  wis: {
    name: 'Мудрость',
    descr: '● *60%* влияет на успешности магии \n● *60%* на вероятность защиты от магии \n● *60%* влияет на лечение \n● Увеличивает количество маны',
  },
  int: {
    name: 'Интелект',
    descr: '● *40%* влияет на успешность магии \n● *40%* на вероятность защиты от магии \n● *40%* влияет на лечение \n● Увеличивает длительность заклинаний',
  },
  con: {
    name: 'Телосложение',
    descr: '● *60%* влияет на вероятность защиты \n● *40%* влияет на востановление энергии \n● Увеличивает запас здоровья',
  },
};
