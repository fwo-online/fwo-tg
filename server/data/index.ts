export * as Harks from './harks';
export * as Profs from './profs';
