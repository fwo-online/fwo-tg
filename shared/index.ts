export * from './character';
export * from './inventory';
export * from './item';
export * from './itemSet';
export * from './shared';
export * from './actions';
export * from './game';
export * from './clan';
