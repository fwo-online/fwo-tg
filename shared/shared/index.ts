export * from './id';
export * from './modifiers';
export * from './player';
export * from './orderSchema';
export * from './message';
export * from './attributes';
export * from './action';
