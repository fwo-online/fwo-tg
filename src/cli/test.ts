import { $ } from 'bun';

async function test() {
  try {
    await $`docker-compose up -d --build db`;

    await $`bun test --preload ./test.setup.ts`;

    await $`docker-compose stop db`;
    process.exit(0);
  } catch {
    process.exit(1);
  }
}

void test();
