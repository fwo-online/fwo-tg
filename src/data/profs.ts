import type { Harks } from '.';

export const profsList = ['m', 'w', 'p', 'l'] as const;

export type Prof = typeof profsList[number];

export type ProfsLvl = Partial<Record<Prof, number>>;

type ProfItem = {
  hark: Harks.HarksLvl;
  descr: string;
  name: string;
  icon: string;
  mag?: Record<string, number>;
}

export const profsData: Record<Prof, ProfItem> = {
  w: {
    hark: {
      str: 10, dex: 8, int: 3, wis: 3, con: 6,
    },
    descr: '—Å—Ç—Ä–æ–Ω–≥',
    name: '–í–æ–∏–Ω',
    icon: 'üõ°',
  },
  l: {
    hark: {
      str: 3, dex: 10, int: 8, wis: 3, con: 6,
    },
    descr: '—Ä–µ–∑–∫–∏–π',
    name: '–õ—É—á–Ω–∏–∫',
    icon: 'üèπ',
  },
  m: {
    hark: {
      str: 3, dex: 3, int: 8, wis: 10, con: 6,
    },
    mag: {
      lightHeal: 1,
    },
    descr: '–≤–æ–ª—à–µ–±–Ω—ã–π',
    name: '–ú–∞–≥',
    icon: 'üîÆ',
  },
  p: {
    hark: {
      str: 3, dex: 3, int: 10, wis: 8, con: 6,
    },
    mag: {
      lightHeal: 1,
    },
    descr: '—Ö–∏–ª–∏—Ç',
    name: '–ñ—Ä–µ—Ü',
    icon: '‚ô±',
  },
};
