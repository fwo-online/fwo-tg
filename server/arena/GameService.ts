import { createGame } from '@/api/game';
import type { Game } from '@/models/game';
import type { LongItem } from '@/arena/Constuructors/LongMagicConstructor';
import { engine } from '@/arena/EngineService';
import { HistoryService, type HistoryItem } from '@/arena/HistoryService';
import type * as magics from '@/arena/magics';
import OrderService from '@/arena/OrderService';
import PlayersService, { type Player } from '@/arena/PlayersService';
import { RoundService, RoundStatus } from '@/arena/RoundService';
import arena from '@/arena';
import { mapValues } from 'es-toolkit';
import EventEmitter from 'node:events';
import { type GameStatus, reservedClanName } from '@fwo/shared';
import { getRandomComponent } from '@/utils/getRandomComponent';

export type KickReason = 'afk' | 'run';

export interface GlobalFlags {
  isEclipsed: { initiator: Player }[];
}

/**
 * GameService
 *
 * @description Обработка около игровой логики
 * @module Service/Game
 * @todo сейчас после того как Player отключился, socket выходит из room.
 * Нужен механизм подключения обратно, если клиент "обновил" страницу или
 * переподключился к игре после disconnect(разрыв соединения)
 */

/**
 * Класс для объекта игры
 */
export default class GameService extends EventEmitter<{
  end: [
    {
      reason: string | undefined;
      statistic: Partial<Record<string, { exp: number; gold: number; nick: string }[]>>;
    },
  ];
  startOrders: [];
  endOrders: [];
  preKick: [{ reason: string; player: Player }];
  kick: [{ reason: string; player: Player }];
  startRound: [{ round: number; status: Record<string, GameStatus[]> }];
  endRound: [{ dead: Player[]; log: HistoryItem[] }];
}> {
  players: PlayersService;
  orders: OrderService;
  round = new RoundService();
  history = new HistoryService();
  longActions: Partial<Record<keyof typeof magics, LongItem[]>> = {};
  info!: Game;
  flags: {
    noDamageRound: number;
    global: GlobalFlags;
  };

  /**
   * Конструктор объекта игры
   * @param players массив игроков
   */
  constructor(players: string[]) {
    super();

    this.players = new PlayersService(players);
    this.orders = new OrderService(this.players, this.round);
    this.flags = {
      noDamageRound: 0,
      global: {
        isEclipsed: [],
      },
    };
  }

  /**
   * Функция проверки окончания игры
   */
  get isGameEnd(): boolean {
    return (
      this.isTeamWin ||
      this.players.alivePlayers.length === 0 ||
      this.flags.noDamageRound > 2 ||
      this.round.count > 9
    );
  }

  get isTeamWin(): boolean {
    const { [reservedClanName]: noClan, ...groupByClan } = this.players.groupByClan(
      this.players.alivePlayers,
    );
    if (!noClan?.length) {
      return Object.keys(groupByClan).length === 1;
    }
    return noClan.length === 1 && !Object.keys(groupByClan).length;
  }

  getEndGameReason() {
    if (this.flags.noDamageRound > 2) {
      return 'noDamageRound';
    }
  }

  get checkRoundDamage(): boolean {
    return !!this.history.hasDamageForRound(this.round.count);
  }

  /**
   * Предзагрузка игры
   */
  preLoading(): void {
    this.initHandlers();
    this.startGame();

    arena.games[this.info.id] = this;

    this.info.players.forEach((playerId) => {
      arena.characters[playerId].gameId = this.info.id;
    });
    // @todo add statistic +1 game for all players
  }

  /**
   * Старт игры
   */
  startGame(): void {
    console.debug('GC debug:: startGame', 'gameId:', this.info.id);
    this.round.initRound();
  }

  /**
   *@todo Остановка игры
   */
  pauseGame(): void {
    console.debug(this.info.id);
  }

  /**
   * Прекик, помечаем что пользователь не выполнил заказ и дальше будет выброшен
   * @param id id игрока, который будет помочен как бездействующий
   * @param reason строка, подставляющаяся в флаг isKicked
   */
  preKick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: preKick', id, 'no player');
      return;
    }
    player.preKick(reason);
    this.emit('preKick', { reason, player });
  }

  /**
   * Функция "выброса игрока" из игры без сохранения накопленных статов
   * @param id id игрока, который будет выброшен
   * @param reason причина кика
   */
  kick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: kick', id, 'no player');
      return;
    }

    this.emit('kick', { reason, player });

    const char = arena.characters[id];
    char.addGameStat({ runs: 1 });
    void char.saveToDb();
    char.autoreg = false;
    this.players.kick(id);
    this.info.players.splice(this.info.players.indexOf(id), 1);
    this.cleanLongMagics();
    char.gameId = '';
  }

  checkRun(player: Player) {
    if (player.flags.isKicked === 'run') {
      this.kick(player.id, player.flags.isKicked);
      return;
    }
  }

  /**
   * Проверяем делал ли игрок заказ. Помечает isKicked, если нет
   * @param player
   */
  checkOrders(player: Player): void {
    if (!player.alive || this.round.count === 1) {
      return;
    }

    if (!this.orders.checkPlayerOrderLastRound(player.id)) {
      if (player.flags.isKicked === 'afk') {
        this.kick(player.id, 'afk');
      } else {
        this.preKick(player.id, 'afk');
      }
    } else {
      /** @todo create clear kick flag method */
      player.preKick(undefined);
    }
  }

  /**
   * Ставим флаги, влияющие на окончание игры
   */
  handleEndGameFlags(): void {
    if (this.checkRoundDamage) {
      this.flags.noDamageRound = 0;
    } else {
      this.flags.noDamageRound += 1;
    }
  }

  recordOrderResult(item: HistoryItem) {
    this.history.addHistoryForRound(item, this.round.count);
  }

  getLastRoundResults() {
    return this.history.getHistoryForRound(this.round.count - 1);
  }

  getRoundResults() {
    return this.history.getHistoryForRound(this.round.count);
  }

  /**
   * @description Завершение игры
   *
   */
  endGame(): void {
    console.log('GC debug:: endGame', this.info.id);
    // Отправляем статистику
    setTimeout(async () => {
      const statistic = this.statistic();
      await this.saveGame();
      this.forAllPlayers((player: Player) => {
        arena.characters[player.id].gameId = '';
      });

      this.emit('end', { reason: this.getEndGameReason(), statistic });
      this.removeAllListeners();

      arena.mm.cancel();

      // FIXME move to client side
      this.forAllPlayers((player: Player) => {
        const char = arena.characters[player.id];
        if (char.expEarnedToday >= char.expLimitToday) {
          char.autoreg = false;
        }
        if (!char.autoreg) return;
        arena.mm.push({
          id: player.id,
          psr: 1000,
          startTime: Date.now(),
        });
      });
    }, 5000);
  }

  /**
   * Создание объекта в базе // потребуется для ведения истории
   * @return Объект созданный в базе
   */
  async createGame() {
    try {
      const dbGame = await createGame(this.players.init);
      this.info = dbGame;
      this.info.id = this.info._id.toString();
      return this;
    } catch (e) {
      console.log('GC debug:: createGame', e);
    } finally {
      this.preLoading();
    }
  }

  /**
   * Очищаем глобальные флаги в бою
   * затмение, бунт богов, и т.п
   */
  refreshRoundFlags(): void {
    this.flags.global.isEclipsed = [];
  }

  /**
   * Подвес
   */
  initHandlers(): void {
    // Обработка сообщений от Round Module
    this.round.subscribe(({ state, round }) => {
      switch (state) {
        case RoundStatus.START_ROUND: {
          this.forAllPlayers(this.checkOrders);
          this.sendStatus(round);
          break;
        }
        case RoundStatus.END_ROUND: {
          this.forAllPlayers(this.checkRun);
          this.emit('endRound', { dead: this.sortDead(), log: this.getRoundResults() });
          this.players.reset();
          this.orders.reset();
          this.handleEndGameFlags();
          if (this.isGameEnd) {
            this.round.unsubscribe();
            this.endGame();
          } else {
            this.refreshRoundFlags();
            this.round.nextRound();
          }
          break;
        }
        case RoundStatus.ENGINE: {
          engine(this);
          break;
        }
        case RoundStatus.START_ORDERS: {
          this.emit('startOrders');
          break;
        }
        case RoundStatus.END_ORDERS: {
          this.emit('endOrders');
          break;
        }
        default: {
          console.log('InitHandler:', state, 'undef event');
        }
      }
    });
  }

  /**
   * Метод сохраняющий накопленную статистику игроков в базу и сharObj
   * @todo нужен общий метод сохраняющий всю статистику
   */
  async saveGame() {
    try {
      const promises = this.info.players.map(async (id) => {
        const player = this.players.getById(id);
        if (!player) {
          return;
        }
        const isWinner = player.alive;

        await arena.characters[id].resources.addResources({
          gold: player.stats.collect.gold,
          exp: player.stats.collect.exp,
          components: player.stats.collect.component
            ? { [player.stats.collect.component]: 1 }
            : undefined,
        });

        const kills = this.players.getKills(id).length;
        const death = isWinner ? 0 : 1;

        arena.characters[id].addGameStat({
          games: 1,
          death,
          kills,
        });
        return arena.characters[id].saveToDb();
      });

      await Promise.all(promises);
    } catch (e) {
      console.log('Game:', e);
    }
  }

  /**
   * Функция послематчевой статистики
   * @todo надо как-то переделать функцию, потому что непонятно, что она мутирует данные
   * @return возвращает строку статистики по всем игрокам
   */
  statistic() {
    this.giveGoldforKill();
    const winners = this.players.alivePlayers;
    const gold = this.players.deadPlayers.length ? 5 : 1;
    winners.forEach((p) => {
      p.stats.addGold(gold);
      p.stats.addComponent(getRandomComponent(60));
    });

    this.players.deadPlayers.forEach((p) => {
      p.stats.addComponent(getRandomComponent(20));
    });

    const playersByClan = this.players.groupByClan();

    const getStatisticString = (p: Player) => ({
      nick: p.nick,
      exp: p.stats.collect.exp,
      gold: p.stats.collect.gold,
      component: p.stats.collect.component,
    });

    const statisticByClan = mapValues(playersByClan, (players) => {
      return players?.map(getStatisticString);
    });

    return statisticByClan;
  }

  /**
   * Функция пробегает всех убитых и раздает золото убийцам
   */
  giveGoldforKill(): void {
    this.players.deadPlayers.forEach((p) => {
      const killer = this.players.getById(p.getKiller());
      if (killer && killer.id !== p.id) {
        killer.stats.addGold(5 * p.lvl);
      }
    });
  }

  /**
   * Функция выставляет "смерть" для игроков имеющих hp < 0;
   * Отсылает сообщение о смерти игрока в последнем раунде
   * @todo сообщение о смерти как-то нормально нужно сделать,
   * чтобы выводило от чего и от кого умер игрок
   */
  sortDead() {
    const dead = this.players.sortDead();
    this.cleanLongMagics();
    return dead;
  }

  /**
   * Очистка массива длительных магий от умерших
   */
  cleanLongMagics(): void {
    this.longActions = mapValues(this.longActions, (longMagicType) => {
      return longMagicType?.filter((act) => {
        return this.players.getById(act.target)?.alive;
      });
    });
  }

  /**
   * Интерфейс для работы со всеми игроками в игре
   * @param f функция применяющая ко всем игрокам в игре
   */
  forAllPlayers(f: (player: Player) => void): void {
    this.players.players.forEach((p) => f.call(this, p));
  }

  getStatus() {
    const playersByClan = this.players.groupByClan();

    return mapValues(playersByClan, (players = []) => {
      return players.map((p) => p.getStatus());
    });
  }

  /**
   * Рассылка состояний живым игрокам
   */
  sendStatus(round: number): void {
    const status = this.getStatus();

    this.emit('startRound', { round, status });
  }
}
