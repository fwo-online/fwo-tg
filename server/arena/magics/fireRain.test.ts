import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { CharacterClass } from '@fwo/shared';
import { times } from 'lodash';
import GameService from '@/arena/GameService';
import { profsData } from '@/data/profs';
import TestUtils from '@/utils/testUtils';
import fireRain from './fireRain';

// npm t server/arena/magics/fireRain.test.ts

describe('fireRain', () => {
  let game: GameService;

  beforeEach(async () => {
    const harks = { ...profsData.m.hark, wis: 20 };
    const initiator = await TestUtils.createCharacter({
      prof: CharacterClass.Mage,
      magics: { fireRain: 3 },
      harks,
    });
    const chars = await Promise.all(times(10, () => TestUtils.createCharacter({})));

    await TestUtils.createClan(chars[0], {
      players: chars.slice(0, 6),
    });

    game = new GameService([initiator.id, ...chars.map(({ id }) => id)]);
    game.players.players.forEach((player, index) => {
      player.resists.fire = index % 3 ? 1 : 0.75;
    });
  });

  beforeEach(() => {
    TestUtils.mockRandom(0.15);
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('should hit clan targets', () => {
    game.players.players[0].proc = 1;

    fireRain.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should hit single target if target has no clan', () => {
    game.players.players[0].proc = 1;

    fireRain.cast(game.players.players[0], game.players.players[8], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
