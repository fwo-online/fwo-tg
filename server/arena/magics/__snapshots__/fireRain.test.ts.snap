// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fireRain should hit clan targets 1`] = `
[
  8,
  2.59,
  2.59,
  3.93,
  2.63,
  2.59,
  3.97,
  8,
  8,
  8,
  8,
]
`;

exports[`fireRain should hit clan targets 2`] = `234`;

exports[`fireRain should hit clan targets 3`] = `
"*asperiores* сотворил _Огненный дождь_ на *magni* нанеся 28.32 урона
\\[ magni 🔥 💔-4.03/2.59 📖40
labore 🔥 💔-5.41/2.59 📖43
iure 🔥 💔-4.07/3.93 📖33
recusandae 🔥 💔-5.37/2.63 📖43
ipsam 🔥 💔-5.41/2.59 📖43
dignissimos 🔥 💔-4.03/3.97 📖32 ]"
`;

exports[`fireRain should hit single target if target has no clan 1`] = `
[
  8,
  8,
  8,
  8,
  8,
  8,
  8,
  8,
  2.57,
  8,
  8,
]
`;

exports[`fireRain should hit single target if target has no clan 2`] = `51`;

exports[`fireRain should hit single target if target has no clan 3`] = `
"*explicabo* сотворил _Огненный дождь_ на *aperiam* нанеся 5.43 урона
\\[ aperiam 🔥 💔-5.43/2.57 📖51 ]"
`;
